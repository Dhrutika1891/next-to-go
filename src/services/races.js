import axios from "axios";

export const getRacesData = async () => {
  const { data: response } = await axios.get(
    "https://api.neds.com.au/rest/v1/racing/?method=nextraces&count=10"
  );
  return response.data;
};
