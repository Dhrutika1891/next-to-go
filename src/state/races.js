import { map, isEmpty, sortBy } from "lodash";

import { setStatus } from ".";
import { getRacesData } from "../services/races";

export const setRaces = (racesPageData) => ({
  type: "set-races",
  racesPageData,
});

const transformRaces = (races) => {
  return (
    !isEmpty(races) &&
    map(sortBy(races.race_summaries, "advertised_start.seconds"), (item) => ({
      ...item,
      isExpired: false,
    }))
  );
};

export const getRaces = () => {
  return async (dispatch) => {
    dispatch(setStatus("LOADING"));

    try {
      const races = transformRaces(await getRacesData());

      const racesPageData = {
        races,
      };

      if (racesPageData) {
        dispatch(setRaces(racesPageData));
        dispatch(setStatus("SUCCESS"));
      } else {
        dispatch(setStatus("EMPTY"));
      }
    } catch (error) {
      dispatch(setStatus("ERROR"));
    }
  };
};

export const expireRaces = (expiredRaceIds = []) => {
  return async (dispatch, getState) => {
    const { racesPageData } = getState();

    dispatch(
      setRaces({
        ...racesPageData,
        races: racesPageData.races.map((race) =>
          expiredRaceIds.indexOf(race.race_id) === -1
            ? race
            : { ...race, isExpired: true }
        ),
      })
    );
  };
};
