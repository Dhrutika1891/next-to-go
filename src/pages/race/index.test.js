import { render, waitFor } from "@testing-library/react";
import React from "react";

import Race from ".";
import { getRacesData } from "../../services/races";

jest.mock("../../services/races", () => {
  return {
    getRacesData: jest.fn(),
  };
});

describe("component race row", () => {
  const mockRacesResponse = {
    next_to_go_ids: [],
    race_summaries: {
      [`e275f2c4-0ee3-47b0-bed3-27d27e15985f`]: {
        advertised_start: { seconds: 1614429960 },
        category_id: "9daef0d7-bf3c-4f50-921d-8e818c60fe61",
        meeting_id: "514af472-4d7e-4c0e-82d7-2f93e96af16c",
        meeting_name: "Romford Bags",
        race_form: {},
        race_id: "e275f2c4-0ee3-47b0-bed3-27d27e15985f",
        race_name: "400Mtrs (A6)",
        race_number: 11,
        venue_country: "UK",
        venue_id: "559f865c-f6e5-47e0-8881-fe6c92f4a508",
        venue_name: "Romford Bags",
        venue_state: "UK",
      },
      [`f9f14b20-aac5-4d29-8903-70618ae8f46d`]: {
        advertised_start: { seconds: 1614429120 },
        category_id: "9daef0d7-bf3c-4f50-921d-8e818c60fe61",
        meeting_id: "fb2eb7ac-77b7-47dd-bd09-b6accb9db48c",
        meeting_name: "Cannington",
        race_form: {},
        race_id: "f9f14b20-aac5-4d29-8903-70618ae8f46d",
        race_name: "Wednesday $20 Burger Beer Bet",
        race_number: 5,
        venue_country: "AUS",
        venue_id: "9b3705d7-d2f4-43ce-97ba-5dde4d3bce77",
        venue_name: "Cannington",
        venue_state: "WA",
      },
    },
  };

  const mockTransformedRaces = [
    {
      advertised_start: {
        seconds: 1614429120,
      },
      category_id: "9daef0d7-bf3c-4f50-921d-8e818c60fe61",
      isExpired: false,
      meeting_id: "fb2eb7ac-77b7-47dd-bd09-b6accb9db48c",
      meeting_name: "Cannington",
      race_form: {},
      race_id: "f9f14b20-aac5-4d29-8903-70618ae8f46d",
      race_name: "Wednesday $20 Burger Beer Bet",
      race_number: 5,
      venue_country: "AUS",
      venue_id: "9b3705d7-d2f4-43ce-97ba-5dde4d3bce77",
      venue_name: "Cannington",
      venue_state: "WA",
    },
    {
      advertised_start: {
        seconds: 1614429960,
      },
      category_id: "9daef0d7-bf3c-4f50-921d-8e818c60fe61",
      isExpired: false,
      meeting_id: "514af472-4d7e-4c0e-82d7-2f93e96af16c",
      meeting_name: "Romford Bags",
      race_form: {},
      race_id: "e275f2c4-0ee3-47b0-bed3-27d27e15985f",
      race_name: "400Mtrs (A6)",
      race_number: 11,
      venue_country: "UK",
      venue_id: "559f865c-f6e5-47e0-8881-fe6c92f4a508",
      venue_name: "Romford Bags",
      venue_state: "UK",
    },
  ];

  const racesPageData = {
    races: mockTransformedRaces,
  };

  const mockState = {
    status: "SUCCESS",
    racesPageData,
  };

  it("should render race page with all element", async () => {
    expect.hasAssertions();

    getRacesData.mockResolvedValue(mockRacesResponse);

    const { getByTestId, getByText, baseElement } = render(
      <Race reducer={mockState} />
    );

    const pageStatus = getByTestId("page-status");
    expect(pageStatus).toBeInTheDocument();

    const loadingStatus = getByTestId("status-loading");
    expect(loadingStatus).toBeInTheDocument();
    expect(loadingStatus).toHaveTextContent("Loading");

    expect(baseElement).toMatchSnapshot();

    await waitFor(() => {
      const racePage = getByTestId("race-page");
      const header = getByText("NEXT TO GO");
      const raceCategories = getByTestId("radio-group");
      const raceTable = getByTestId("race-table");

      expect(racePage).toBeInTheDocument();
      expect(header).toBeInTheDocument();
      expect(raceCategories).toBeInTheDocument();
      expect(raceTable).toBeInTheDocument();
    });
  });
});
