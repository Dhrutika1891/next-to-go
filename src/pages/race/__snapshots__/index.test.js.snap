// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component race row should render race page with all element 1`] = `
<body>
  <div>
    <div
      class="page-status"
      data-testid="page-status"
    >
      <div
        data-testid="status-loading"
      >
        <div
          class="spinner-border text-primary"
        />
        <div>
          Loading...
        </div>
      </div>
    </div>
  </div>
</body>
`;
